### ISSUE TITLE
Explain how to specify a subprotocol

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-thing-description#2217]
w3c/wot-binding-templates#249 moved subprotocols back from the appendix however the current bindings do not explain that clearly. The CoAP binding does it at https://w3c.github.io/wot-binding-templates/bindings/protocols/coap/index.html#observing-resources but not under a special section.
We should update the template as well.

The overall mechanism is also a bit fuzzy, further discussion is needed. We should specify what a subprotocol can specify and what it cannot.

Another alternative would be to remove the subprotocol mechanism and make everything more descriptive (more vocabulary items).

Something we really should focus on (with subprotocol or not) is how to use WebSockets in TDs. This needs a separate discussion which can involve more redesign. Also how do we handle MQTT over WS? Are there other protocols that do this? How to handle RPC in general?

Coming from call of 22.02.

**Comment by egekorkan** [comment]
We need to bring this discussion back as it impacts the registry mechanism and implementation. A subprotocol can be dependent on the supprotocol (if that is a term :)) In https://github.com/w3c/wot-binding-templates/issues/361 this is also a bit of a concern. The word `subprotocol` seems to be only used in the context of WebSockets outside of WoT. Also we do not explain this properly anywhere? How does it differ from the `+` notation in URI schemes? Should one use `mqtt+ws://` or `ws://` with `subprotocol:mqtt`? With Modbus, we have `modbus+tcp` and `modbus+rtu` for example. Also there is `opc.tcp` (see [here](https://reference.opcfoundation.org/specs/OPC-10000-6/7)). Who is using what notation for which meaning?

Maybe this is a long discussion or a philosophical one but we need to start it and add the correct guidelines.

**Comment by benfrancis** [comment]
@egekorkan wrote:
> We need to bring this discussion back as it impacts the registry mechanism and implementation.
...
> The word subprotocol seems to be only used in the context of WebSockets outside of WoT. Also we do not explain this properly anywhere?

You are right that the term "sub-protocol" is only really formally defined in WebSockets, but I think most developers would intuitively understand what you meant if you used it with another protocol. There is an [existing definition](https://w3c.github.io/wot-thing-description/#subprotocols) in section 8.1.1.1 of the WoT Thing Description specification which seems reasonable, but the values are currently largely unconstrained.

I don't think the term "sub-protocol" is commonly used with HTTP, although countless examples exist. They are sometimes called "domain application protocols" or "HTTP-based application layer prototocols" but so far "sub-protocol" is the best general purpose term I've been able to come up with.

> In https://github.com/w3c/wot-binding-templates/issues/361 this is also a bit of a concern.

Arguably SSE is not really a sub-protocol, it's a MIME type (`text/event-stream`) for streaming text over HTTP. The specification does define one optional new HTTP header (`Last-Event-ID`) but all the others (`Accept: text/event-stream`, `Connection: keep-alive`, `Cache-Control: no-cache` etc.) are just standard HTTP. Most of the EventSource specification is dedicated to defining a JavaScript API for consuming a stream of server-sent events rather than a protocol.

Having said that, it is quite convenient to be able to use `sse` as a subprotocol of HTTP with an implied `text/event-stream` content type in the case where you are wrapping another content type in events. E.g. setting subprotocol as `sse` and `Content-type` as `application/json`. Maybe this could be specified in the HTTP protocol binding.

I think it's worth re-visiting the examples given in the specification (`longpoll`, `websub` and `sse`) because none of them are particularly well defined. Perhaps each protocol binding should explain whether the protocol supports sub-protocols and if so what the mechanism is and how they are defined. The TD specification actually already says "The subprotocols can be defined and explained as a part of a protocol [binding specification](https://w3c.github.io/wot-thing-description/#dfn-binding-specification)."

> Should one use mqtt+ws:// or ws:// with subprotocol:mqtt

In this case you should use ws:// because that's what the WebSocket specification says, but the sub-protocol mechanism differs between protocols.

> With Modbus, we have modbus+tcp and modbus+rtu

I think this is a bit different because Modbus is the application protocol and the part after the + is defining the transport (Note these are also unofficial URI schemes not registered with IANA).

In conclusion I think WoT's current sub-protocol definition is a good starting point but it could probably be tightened up a little, by outlining how protocol bindings define a sub-protocol mechanism and set of valid values, and what those sub-protocols can and can't constrain.

**Comment by benfrancis** [comment]
(obviously a key new use case for this term is the new WoT Thing Protocol deliverables, so that is a good opportunity to refine this mechanism)

**Comment by egekorkan** [comment]
Thanks @benfrancis ! I kind of agree with you but see a different reality, sadly... For example CoAP defines in IANA with permanent status (https://www.iana.org/assignments/uri-schemes) 6 URI schemes. coap+ws is one such scheme. There is also a CoAP subprotocol registered at https://www.iana.org/assignments/websocket/websocket.xml 

While Modbus (and OPC UA too) are not registering their URI schemes in IANA, they are using the same pattern as CoAP. OPC UA does define subprotocol names at https://reference.opcfoundation.org/specs/OPC-10000-6/7.5 (7.5.2 is the right link but this provides more context around it). Talking with Cris, we have the following:

- A TD parser chooses the binding based on what it sees in the form. If they need to look into two keywords to choose a binding, it becomes more cumbersome. In MQTT.js, you are expected to put a `ws://` URI into the broker field.
- It is possible to take advantage of the uri scheme as much as we want. There is no limitation to use multiple `+` or `.`.
- Doing a step by step parsing of the URI Scheme makes sense. In `mqtt+ws` case, the parser first detects mqtt and realizes that it needs to choose an mqtt driver. Then gives the URI in the format their driver expects (`ws://` or else). 

My point of view after thinking a bit more: Even if it is a websocket subprotocol, you always need a library on top of it. So I would say that there is no difference between TCP and WS in this case. No one uses a TCP library to make mqtt messages from scratch. The only difference is for browser implementations to detect that this is not a MQTT over TCP Thing but WS, so I can communicate with it. Thinking of this, step by step parsing is not a good idea as MQTT would not be considered to be usable even though that ws is behind. Also for node-wot (@danielpeintner @relu91), I am thinking that we should think differently for websockets. MQTT binding is not packaged into the browser bundle but it should be for mqtt+ws use case. If we move all those websocket compatible implementations to the websocket package, that would be solved but creates duplicate code?

**Comment by benfrancis** [comment]
@egekorkan wrote:
> CoAP defines in IANA with permanent status (https://www.iana.org/assignments/uri-schemes) 6 URI schemes. coap+ws is one such scheme. There is also a CoAP subprotocol registered at https://www.iana.org/assignments/websocket/websocket.xml

It's a bit of a weird case but I think these two registrations are expressing the same thing from two different perspectives:
1. Different URI schemes for CoAP as a protocol using different transports (TCP and WebSockets instead of the default UDP)
2. A sub-protocol registration for CoAP as a sub-protocol of the WebSocket protocol (for use in sub-protocol negotiation when opening a WebSocket connection)

> While Modbus (and OPC UA too) are not registering their URI schemes in IANA, they are using the same pattern as CoAP. 

Yes, they are using the former pattern, because Modbus and OPC UA are protocols that support different transports. You wouldn't say that Modbus is a sub-protocol of TCP. The application protocol is Modbus and it's using TCP as a transport.

> A TD parser chooses the binding based on what it sees in the form. If they need to look into two keywords to choose a binding, it becomes more cumbersome.

I think it's simpler to choose a protocol binding based on the URI scheme, then the binding can define how sub-protocols work if they are supported.

> It is possible to take advantage of the uri scheme as much as we want. There is no limitation to use multiple + or .

Except that I wouldn't assume all WebSocket clients would support a coap+ws:// URI, but they must support ws:// if they conform to the WebSocket specification. The WebSocket specification defines a specific in-protocol mechanism for negotiating a sub-protocol when opening a connection, and that doesn't involve special URI schemes. So in the case of WebSockets in particular I think it makes sense for the sub-protocol to be included in the Thing Description as a separate member of the Form.

> Doing a step by step parsing of the URI Scheme makes sense. In mqtt+ws case, the parser first detects mqtt and realizes that it needs to choose an mqtt driver. Then gives the URI in the format their driver expects (ws:// or else).

Just because a Consumer supports MQTT doesn't mean it supports MQTT over WebSockets. And just because it supports WebSockets doesn't mean it supports MQTT over WebSockets. They are two separate pieces of information needed to select a Form. A WebSocket endpoint should really use the ws:// URI scheme and the WebSocket's built-in sub-protocol negotiation, but if you want to support special URI schemes like mqtt+ws then I guess that URI scheme could map onto the MQTT protocol binding and be dealt with there. i.e. multiple URI schemes can map onto the same protocol binding rather than requiring pre-processing of all URIs.

As another example, I don't think we would want to define new wtp+http and wtp+ws URI schemes for the WoT Thing Protocol, because they wouldn't be directly supported by the vast array of different HTTP and WebSocket clients out there - which is the whole point of building on top of existing protocols in the first place. Requiring a Consumer to carry out an extra step of transforming the URI scheme seems over-complicated to me. (It would be nice to create a new dedicated wtp:// URI scheme rather than dealing with the messy concept of HTTP sub-protocols, but again it wouldn't be supported by existing clients, which is the whole point. I assume this is why other HTTP "sub-protocols" like ActivityPub, gRPC, MCP and Solid/LWS don't have their own URI schemes).

> Even if it is a websocket subprotocol, you always need a library on top of it. So I would say that there is no difference between TCP and WS in this case.

I understand your point, but WebSocket is still an application layer protocol built on top of TCP and HTTP which defines its own sub-protocol negotiation mechanism, it is not equivalent to TCP (which is not an application layer protocol and does not have its own URI scheme).

I think for WebSockets the current solution of using the `ws://` URI scheme and a separate `subprotocol` member makes sense. For other protocols it is less clear because the concept of sub-protocols is less clearly defined, but each protocol binding can explain whether and how sub-protocols can be used, and perhaps multiple URI schemes can map onto the same protocol binding if necessary.

**Comment by egekorkan** [comment]
TD call today:
- We should list a reasonable set of these combinations. IoT protocols, WoT Thing protocol, OT protocols. @egekorkan  to write a list in this issue thread. The approach (new or the current one) should fit those common cases.
- `subprotocol` can be deprecated if we allow registering new URI Schemes via the registry but that can get pushback. Even for `longpoll+http://`
- What should be registered in the registry? All possible URI schemes for this binding? If we say other terms are allowed, those need to be registered as well. For CoAP, one needs to register 6 schemes now (coap, coap+ws
- long poll needs to be thought. Is it a convention that simply goes in HTTP binding?

**Comment by benfrancis** [comment]
> We should list a reasonable set of these combinations. IoT protocols, WoT Thing protocol, OT protocols. @egekorkan to write a list in this issue thread. The approach (new or the current one) should fit those common cases.

This is a good idea, but I would suggest being careful not to use an interpretation of "sub-protocol" which is too broad. A sub-protocol is only needed where there is an existing established application layer protocol and additional semantics are added on top (like WebSocket sub-protocols, or domain application protocols built on top of HTTP).

For example, Modbus and OPC-UA are not "sub-protocols" of TCP, TCP is just the transport protocol for those application protocols.

> subprotocol can be deprecated if we allow registering new URI Schemes via the registry but that can get pushback. Even for longpoll+http://

Do you mean we would define made up URI schemes in the WoT registry (but not the IANA registry) in order to include a subprotocol identifier in the URI scheme, like wtp+http:// and wtp+ws:// ? I would strongly recommend against that. URLs have been a fundamental building block of the World Wide Web for 35 years with an established IETF registry of reserved URI schemes. We should not invent special URI schemes which have special meanings in WoT but nowhere else. There are good reasons for using existing standardised URI scheme then defining a subprotocol on top, particularly for HTTP which is the de facto narrow waist of the web. 

> What should be registered in the registry? All possible URI schemes for this binding?

Yes, I think so. We already need to map multiple URI schemes onto one protocol binding (e.g. http and https or ws and wss). If there are other URI schemes that people use with that binding then it would be a good idea to include them.

coap+ws:// is a bit of an unusual example because there is also an IANA registration for a coap WebSocket sub-protocol. One way to approach that could be that the coap+ws:// URI scheme maps onto the CoAP protocol binding (with an explanation in the binding document of what that means), but a WebSocket binding also refers to the WebSocket sub-protocol registry which includes the coap sub-protocol, so both approaches can be used.

> long poll needs to be thought. Is it a convention that simply goes in HTTP binding?

I agree that longpoll is not really a sub-protocol, just a communication pattern which uses standard HTTP requests and responses.
